src/parser.cpp

```diff
@@ -121,24 +121,42 @@ static AstNode *ast_parse_container_decl_auto(ParseContext *pc);
 static AstNode *ast_parse_container_decl_type(ParseContext *pc);
 static AstNode *ast_parse_byte_align(ParseContext *pc);
 
-ATTRIBUTE_PRINTF(3, 4)
+ATTRIBUTE_PRINTF(4, 5)
 ATTRIBUTE_NORETURN
-static void ast_error(ParseContext *pc, Token *token, const char *format, ...) {
+static void ast_error(ParseContext *pc, Token *token, ErrorMsg *note, const char *format, ...) {
```

Comment:
I think it is better to separate error messages from the printing and exiting logic.
We can have:
* `ErrorMsg *ast_error(ParseContext *pc, Token *token, const char *format, ...)`
* `ErrorMsg *ast_error_note(ParseContext *pc, Token *token, const char *format, ...)`
* `void ast_report_and_abort(ParseContext *pc, size_t messages_len, ErrorMsg **messages);`